### PR TITLE
Reload tooltip in the symbol tree also on tag update

### PR DESCRIPTION
**Pull request by techee**
Because function return types are not used to determine tag equality,
we need to also update the tooltip of an existing tag otherwise the return
type doesn't get updated when changed.

Fixes https://github.com/geany/geany/issues/474

### CODE REVIEW

**Comment by techee** [comment]
I was playing with the code and even though I hoped for some better tree creation/update optimization, these patches are about the best I could achieve. I was wrong about some things in my previous comments and after reading the code a bit more, I think in general it's pretty well optimized (I guess one could improve the performance by creating a custom GtkTreeViewModel but that's just too much work).

**Comment by techee** [comment]
When talking about profiles, there's one more hotspot in Geany which often pops up in the profile (and I've seen it again when profiling the performance of a file with many tags). It's the WordList::InList() function of Scintilla. This is called to colorize type names by scintilla passed from Geany here:

https://github.com/geany/geany/blob/master/src/document.c#L2672

This thing slows down editing when there are 

1. either too many tags
2. or not so many tags but the edited file is too big

which means we can't easily set some maximum number of type names in the code and pass only those to Scintilla. Would it be OK to add a Various settings like colorize_type_names where this could be disabled? I think it could help to people who have performance problems with Geany. What do you think?

Apart from this I don't see any other performance problem on Geany side so when someone complains we can just say "sorry, it's a problem in Scintilla, we can't do anything about it" (we have been doing this already but it hasn't always been the case ;-).

**Comment by techee** [comment]
Or maybe we could recolorize the document only when the keywords change (which shouldn't happen when e.g. editing the function body). Will try.

**Comment by elextr** [comment]
> (we have been doing this already but it hasn't always been the case ;-).

Of course it *always* has been Scintilla's fault ;-)

> Or maybe we could recolorize the document only when the keywords change (which shouldn't happen when e.g. editing the function body). Will try.

Try by all means, but as soon as a { is typed, like for an if clause, the function body has just changed, and that erm ... "upsets" the rest of the file :-)  So I doubt it will work well.

Anyhow, colourization is in Scintillas control, so its still its fault ;-)

On limiting the number of names to be colourized, really you should just turn it off, limiting it will give you only some of the symbols, and thats just plain idiotic. Which ones will it get? Those starting with a-p maybe? So "string", the most used symbol ever, won't be colourized, or all those zthingys in tags code ;-).  Or maybe give it every second typename to colourize, that should really help the user ;-)

Just make a pref to turn it off (if there isn't already a pref I missed).

Or don't use a plugin that adds every tag for every file it can manage to find to parse, can't remember its name just now ;-P

**Comment by techee** [comment]
> (we have been doing this already but it hasn't always been the case ;-).
>
> Of course it *always* has been Scintilla's fault ;-)
>
> Or maybe we could recolorize the document only when the keywords change
> (which shouldn't happen when e.g. editing the function body). Will try.
>
> Try by all means, but as soon as a { is typed, like for an if clause, the
> function body has just changed, and that erm ... "upsets" the rest of the
> file :-) So I doubt it will work well.
>
But it will slow-down just typing the single '{' character (and then later
after typing '}'), all the rest will be fast. I can survive 300ms editor
delay after typing { when creating a function but not after every entered
character.

> Anyhow, colourization is in Scintillas control, so its still its fault ;-)
>
> On limiting the number of names to be colourized, really you should just
> turn it off, limiting it will give you only some of the symbols, and thats
> just plain idiotic. Which ones will it get? Those starting with a-p maybe?
> So "string", the most used symbol ever, won't be colourized, or all those
> zthingys in tags code ;-). Or maybe give it every second typename to
> colourize, that should really help the user ;-)
>
So not being able to edit anything is better in your opinion than not
seeing some keywords (in fact just typenames) colorized? Of course the
built in keywords will always be colorized because they are in a different
keyword group (char, int, etc.). What I find idiotic is an "editor" you
cannot use for editing rather than not having some of the types colorized
(this happens already when you don't have a tab open with a file that
contains the tag definition - so I don't really understand the problem).

Anyway, unfortunately I don't see any automatic way to do it so it won't
happen.

> Just make a pref to turn it off (if there isn't already a pref I missed).
>
At the end I think I will try to fix this in Scintilla (when I have time) -
I think Scintilla could sort the keywords and binary-search them which
could help.

However, I think it would be nice to have an option that disables all the
bells and whistles Geany does with tags (tag parsing, tag management, tags
in the sidebar, typename colorization, code completion) so we can see if
something in Geany is the bottleneck. So I have introduced an
"enable_tag_parsing" preferences option which enables/disables tag parsing
globally (and as a result all the rest of the tag stuff because there are
no tags).

> Or don't use a plugin that adds every tag for every file it can manage to
> find to parse, can't remember its name just now ;-P
>
My ProjectOrganizer plugin does it by default just if the number of source
files is reasonably low so it won't slow-down things. Maybe time to try it
to see what it really does instead of making some false assumptions. But
tags may come from other open source files and if the files are big enough,
you can get many tags just from the open files without any additional
plugin.

**Comment by techee** [comment]
I think (hope) I've addressed the comments from the previous review in the updated patches. The extra commit is the enabling/disabling tag parsing which is IMO useful to have both for diagnostic purposes but  maybe also for some very low-end machines (raspberry pi comes to my mind).

**Comment by elextr** [comment]
On 25 April 2015 at 07:32, Jiří Techet <notifications@github.com> wrote:

> > (we have been doing this already but it hasn't always been the case ;-).
> >
> > Of course it *always* has been Scintilla's fault ;-)
> >
> > Or maybe we could recolorize the document only when the keywords change
> > (which shouldn't happen when e.g. editing the function body). Will try.
> >
> > Try by all means, but as soon as a { is typed, like for an if clause, the
> > function body has just changed, and that erm ... "upsets" the rest of the
> > file :-) So I doubt it will work well.
> >
> But it will slow-down just typing the single '{' character (and then later
> after typing '}'), all the rest will be fast. I can survive 300ms editor
> delay after typing { when creating a function but not after every entered
> character.
>

​Yeah, I guess the parser is skipping the rest of the file until the {}s
match again so there would be only one hiccup on { and one on }.​



>
> > Anyhow, colourization is in Scintillas control, so its still its fault
> ;-)
> >
> > On limiting the number of names to be colourized, really you should just
> > turn it off, limiting it will give you only some of the symbols, and
> thats
> > just plain idiotic. Which ones will it get? Those starting with a-p
> maybe?
> > So "string", the most used symbol ever, won't be colourized, or all those
> > zthingys in tags code ;-). Or maybe give it every second typename to
> > colourize, that should really help the user ;-)
> >
> So not being able to edit anything is better in your opinion than not
> seeing some keywords (in fact just typenames) colorized? Of course the
> built in keywords will always be colorized because they are in a different
> keyword group (char, int, etc.). What I find idiotic is an "editor" you
> cannot use for editing rather than not having some of the types colorized
> (this happens already when you don't have a tab open with a file that
> contains the tag definition - so I don't really understand the problem).
>

​Would work fine if the colourization of types is turned off :)​  Which is
what I'm saying, turn it off not half off.



>
> Anyway, unfortunately I don't see any automatic way to do it so it won't
> happen.
>

​Various pref?  Involves the liveware, but better than being uneditable.​



>
> > Just make a pref to turn it off (if there isn't already a pref I missed).
> >
> At the end I think I will try to fix this in Scintilla (when I have time) -
> I think Scintilla could sort the keywords and binary-search them which
> could help.
>

​My quick look at InList() it radix searches the first two bytes, then
​linearly.  Maybe just extend the radix tablets dynamically if the linear
part gets beyond some length.



>
> However, I think it would be nice to have an option that disables all the
> bells and whistles Geany does with tags (tag parsing, tag management, tags
> in the sidebar, typename colorization, code completion) so we can see if
> something in Geany is the bottleneck. So I have introduced an
> "enable_tag_parsing" preferences option which enables/disables tag parsing
> globally (and as a result all the rest of the tag stuff because there are
> no tags).
>
>
​Doesn't setting "menu->preferences->editor->completions->symbol list
update frequency"​ to zero do that?

**Comment by techee** [comment]
> ​Various pref? Involves the liveware, but better than being uneditable.​


I general I try to avoid adding various settings options because

1. Users won't find/enable them
2. The amount of settings knobs might get out of control if everybody adds
something he needs for some reason.

This case isn't something really critical (even with linux-kernel-size
projects the editor is still usable, you just see this function in the
profile) so I'd rather try to find a way to improve the performance in the
code and even if not possible, just keep it the way it is without any extra
pref option.


> >
> ​Doesn't setting "menu->preferences->editor->completions->symbol list
> update frequency"​ to zero do that?


Indeed! I assumed it was just for the sidebar but yes, it seems to skip the
parsing too so this patch isn't needed. Thanks for letting me know about it.

**Comment by techee** [comment]
Repushed again with fixes to the above-discussed issues.

**Comment by b4n** [comment]
As usual when force-pushing we lose the comments on replaced commits… so:

> Yeah, I thought about this too. One possibility could be to have one more dummy entry at the end of the tree with "..." or "(truncated)" but on the other hand users won't see it very often so the question is whether it's worth the work - I don't think there will be any real source file containing so many tags (if we consider something like 20 lines of code per tag, the source would have 40000 lines) - it will be something generated, some huge html or something similar.

Agreed it's probably not very important, but there are some headers with legitimately so many tags. e.g., */usr/include/elf.h* on my Debian has apparently 2458 tags according to ctags, and all those are interesting.  Okay, maybe not so much to browse directly in the tree though.

Also, do we really need that in addition of the other speedup changes here?  Even without any of these, I don't notice slowdowns editing elf.h (maybe higher CPU usage, but not to any point where it'd get in the way of editing itself).  With this branch it seem to use a lot less CPU (which is great!), so goes basically from "totally usable, but more might not" to "what, 2.5k tags you say?  didn't notice".

Rough numbers form top(1) while typing (on one line) at the same pace (fast, but lets the reparsing kick in):
* master, symbol sidebar *not* visible: ~28%
* master, symbol sidebar visible: ~75%
* retval_refresh, symbol sidebar *not* visible: ~17%
* retval_refresh, symbol sidebar visible: ~25%
* retval_refresh (unlimited), symbol sidebar *not* visible: ~14%
* retval_refresh (unlimited), symbol sidebar visible: ~25%

With the file concatenanted twice (~4918 tags):
* master, symbol sidebar *not* visible: ~88% (short locks noticeable)
* master, symbol sidebar visible: ~95% (short locks noticeable)
* retval_refresh, symbol sidebar *not* visible: ~25%
* retval_refresh, symbol sidebar visible: ~30%
* retval_refresh (unlimited), symbol sidebar *not* visible: ~23%
* retval_refresh (unlimited), symbol sidebar visible: ~33%

Okay, those are obviously not very precise, but also obviously suggests that cutting the list doesn't change so much either…

**Comment by techee** [comment]
@b4n After thinking about it maybe we don't need to optimise the editing times - files with so many tags will hardly ever be edited, they are probably generated in some way and are only needed to be viewed. So we should only make sure that the tree generation doesn't slow-down file opening too much. For 10000 tags it takes about 1s to create the tree, for 20000 it's about 4s (and further increases non-linearly) so I'd say those 10000 look like a reasonable number. What do you think?

**Comment by techee** [comment]
OK, one more patch. After this I think Geany scales pretty well:

* for big files (this patch)
* for files with many tags (the sidebar patches)
* for many tagmanager tags (my last year's TM patches)

**Comment by techee** [comment]
The precise value of the "too big" file size (similarly to the "too many" sidebar entries value) is a matter of discussion. I used 10MB where file opening took something like 3s but it can be changed. But IMO there should definitely be some limit because nobody will wait 3 minutes to open a file.

**Comment by elextr** [comment]
Rather than some hard coded majik number why not just add "None" to the list of selectable filetypes on the Open dialog?

**Comment by techee** [comment]
I open files from command-line quite often so there would have to be a command-line switch to do that (and frankly, nobody, including me, would use that). But the main point of the patch is Geany should behave reasonably well by itself no matter what file you try to open and prevent long freezes (e.g. if a user doesn't notice the file is huge and doesn't select None as you suggest).

**Comment by b4n** [comment]
> @b4n After thinking about it maybe we don't need to optimise the editing times - files with so many tags will hardly ever be edited, they are probably generated in some way and are only needed to be viewed. So we should only make sure that the tree generation doesn't slow-down file opening too much. For 10000 tags it takes about 1s to create the tree, for 20000 it's about 4s (and further increases non-linearly) so I'd say those 10000 look like a reasonable number. What do you think?

OK, sounds reasonable.

> The precise value of the "too big" file size (similarly to the "too many" sidebar entries value) is a matter of discussion. I used 10MB where file opening took something like 3s but it can be changed. But IMO there should definitely be some limit because nobody will wait 3 minutes to open a file.

OK, we already talked about a warning for large files in the past.  Not completely sure why it wasn't done, but might only have been missing impl and no clear decision on how it should be done.

---

All this said, I don't like arbitrary limitations very much.  Maybe this could warrant a setting, possibly hidden, only with sensible defaults, but so that someone really working on very huge files with a system powerful enough could still enable it.
At the very least we should have named constants for this defined together somewhere.

Maybe instead of an arbitrary change for large files, we could have a dialog opening up, asking whether the user really meant to open it, warning him it could lead to dramatic slowdowns and letting him chose to disable things that might slow down (e.g. setting filetype to None)?
big-file-dialog
Just some ideas.

**Comment by elextr** [comment]
> OK, we already talked about a warning for large files in the past. Not completely sure why it wasn't done, but might only have been missing impl and no clear decision on how it should be done.

I believe it got mired in the inevitable bikeshed on the size of the number.

My vote is for a setting with 0 to disable, *and* a dialog when the file is bigger than that setting, having the "don't ask again" is a good idea.  That will mitigate the importance of the default value, and I will be happy with anything > 256MB.

But we should still have "None" on the open dialog.  I coulda sworn it was there, but anyway :)

**Comment by techee** [comment]
@b4n The dialog looks good to me. Have you written some code to create the dialog or used some prototyping thing? If you have some code, could you please post it somewhere so I can steel it? :-)

@elextr The 256MB limit is fine when the filetype isn't recognized (so it's None already) but if it is recognized then it must be some much lower value (with 10MB file and 40 characters per line on averagethe the file has 250000 lines at which point Geany becomes pretty unusable already; with 256MB it's 6400000 lines and we are talking about 2 minutes of file opening time and users will just kill Geany at this point).

**Comment by techee** [comment]
Note to self: steal != steel

**Comment by techee** [comment]
I found the sources of the quadratic complexity in the tree code:

1. Basically the tree store behaves as a list so if we append values to it, the whole list has to be walked to the very end. Prepend instead.
2. gtk_tree_store_set() unfortunately calls gtk_tree_store_get_path() internally which also goes through the list to get the index of the updated element.

Fixing (1) means that the tree creation is really fast now (200000 elements in 3s). For this reason I dropped the patch limiting the number of entries in the tree.

Unfortunately the gtk_tree_store_set() problem cannot be fixed easily (without making our own model which isn't worth the work). Instead of updating the items we could remove them and re-add them with the updated values (I tried that) which is much faster than gtk_tree_store_set() but causes some unwanted repositions of the tree view when scrolled and it's stupid to have the tree in the sidebar jumping while editing.

So to sum up the performance:
1. tree creation is fast
2. editing is fast unless a line is added/removed (which causes the gtk_tree_store_set() calls)

**Comment by b4n** [comment]
> @b4n The dialog looks good to me. Have you written some code to create the dialog or used some prototyping thing? If you have some code, could you please post it somewhere so I can steel it? :-)

That's a quick prototype with Glade, there is no real code.  If you want the (hacked) Glade file, I can give it, but it's not very polished.

> […] and I will be happy with anything > 256MB.

Just to be clear, in my mockup "256MB" was a random value for the size of the file being open, not the limit.

> Fixing (1) means that the tree creation is really fast now (200000 elements in 3s). For this reason I dropped the patch limiting the number of entries in the tree.

good! :)  Indeed order shouldn't matter as we sort the result anyway.

> Unfortunately the gtk_tree_store_set() problem cannot be fixed easily (without making our own model which isn't worth the work). Instead of updating the items we could remove them and re-add them with the updated values (I tried that) which is much faster than gtk_tree_store_set() but causes some unwanted repositions of the tree view when scrolled and it's stupid to have the tree in the sidebar jumping while editing.

yeah updating the items and avoiding UI changes was one of the goal when I rewrote the updating code a few years ago, without it the UI changes becomes rapidly annoying when you're editing and see them happen on the side.


@techee looks OK, do you think it is ready for merge?

**Comment by b4n** [comment]
BTW, FTR:

https://developer.gnome.org/gtk3/stable/GtkTreeModel.html
> The lifecycle of an iterator can be a little confusing at first. Iterators are expected to always be valid for as long as the model is unchanged (and doesn’t emit a signal). The model is considered to own all outstanding iterators and nothing needs to be done to free them from the user’s point of view. Additionally, some models guarantee that an iterator is valid for as long as the node it refers to is valid (most notably the GtkTreeStore and GtkListStore). Although generally uninteresting, as one always has to allow for the case where iterators do not persist beyond a signal, some very important performance enhancements were made in the sort model. As a result, the GTK_TREE_MODEL_ITERS_PERSIST flag was added to indicate this behavior.

**Comment by techee** [comment]
@b4n Yep, I think it's ready for merge.

Is the GtkTreeModel part of the documentation related to this case? (I'm not sure I completely understand the docs and when the flag should be used.)

**Comment by b4n** [comment]
the quote is just about the fact that with GtkTreeModel it is alright to prepend, and it won't invalidate iters on rows after it (as the update code uses a lot of iter caching).  that's all I meant to underline (because I know iters are temporary and wondered if this use case was alright)
